homeassistant/components/device_tracker/unifi_direct.py

```diff
@@ -0,0 +1,139 @@
+"""
+Support for Unifi AP direct access.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/device_tracker.unifi_direct/
+"""
+import logging
+import json
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.device_tracker import (
+    DOMAIN, PLATFORM_SCHEMA, DeviceScanner)
+from homeassistant.const import (
+    CONF_HOST, CONF_PASSWORD, CONF_USERNAME,
+    CONF_PORT)
+
+REQUIREMENTS = ['pexpect==4.0.1']
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_SSH_PORT = 22
+UNIFI_COMMAND = 'mca-dump | tr -d "\n"'
+UNIFI_SSID_TABLE = "vap_table"
+UNIFI_CLIENT_TABLE = "sta_table"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_SSH_PORT): cv.port
+})
+
+
+# pylint: disable=unused-argument
+def get_scanner(hass, config):
+    """Validate the configuration and return a Unifi direct scanner."""
+    scanner = UnifiDeviceScanner(config[DOMAIN])
+    if not scanner.connected:
+        return False
+    return scanner
+
+
+class UnifiDeviceScanner(DeviceScanner):
+    """This class queries Unifi wireless access point."""
+
+    def __init__(self, config):
+        """Initialize the scanner."""
+        self.host = config[CONF_HOST]
+        self.username = config[CONF_USERNAME]
+        self.password = config[CONF_PASSWORD]
+        self.port = config[CONF_PORT]
+        self.ssh = None
+        self.connected = False
+        self.last_results = {}
+        self._connect()
+
+    @property
+    def connected(self):
```

Comment:
No need to make this a property. Keep it as a regular instance attribute.

Reply:
The reason I made it a property is because without it, the test_get_device_name fails with the following message.

```
AttributeError: <class 'homeassistant.components.device_tracker.unifi_direct.UnifiDeviceScanner'> does not have the attribute 'connected'
```

Probably something I could get around, but I am not sure how.

Reply:
You should solve that by patching correctly in the test and not by changing the production code, in my opinion. This answer has some suggestions.
https://stackoverflow.com/a/31710001

I'd mock the whole class and then set the attribute on the mock return value with a new mock.

Reply:
Okay, I'll take a look at this. Thanks!

Reply:
An alternative approach could be to patch and autospec the imported pexpect object that is stored in `ssh` on the device scanner. Then you could use more of the production logic in the tests.